template: pull request checklist

vilvo opened the pull request:
Initial pull request template proposal for discussion. 
Checklist is intended to reduce things to remember.
Intent is to support checking with `pre-commit-hooks` checks - where possible.

vilvo commented:
Renders like this in PRs
![screencapture](https://github.com/tiiuae/ghaf/assets/1027150/2f9a9310-4c90-4736-88b4-19eb71e13903)

brianmcgillion reviewed (approved):
Great first start.

mikatammi reviewed (commented):
Should we add some "written fields" to the beginning of the template?

Here's nixpkgs project's template btw https://github.com/NixOS/nixpkgs/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1

I think we could add similar headlines, like "Description of changes".

We could also add headline, "How this was tested?" or "How testers are supposed to test this?"

mikatammi commented:
Many times when I'm looking at PRs, I'm thinking "Why has this PR been made, and what is the expected outcome, how should this thing be used?"

avnik commented:
I'd like to add `nix flake check` to checklist (and integrate devour-flake, to check that all outputs of flake evaluable)

mikatammi commented:
Flake check currrently does not work: https://github.com/tiiuae/ghaf/issues/296

brianmcgillion commented:
> I'd like to add `nix flake check` to checklist (and integrate devour-flake, to check that all outputs of flake evaluable)

Being worked on at the moment.

brianmcgillion commented:
```
nix flake check --accept-flake-config --allow-import-from-derivation

ghaf on flake-parts [$] via  impure (Ghaf-devshell-env) took 40m48s 
```


a clean bill of health for nix flake check we still need to address the IFD issues, (@mikatammi has raised items about 2 so far),  in some derivations but at least the whole thing goes through now.

it needs a fairly beefy machine as it ran at about 35GB ram once it got going also needs to have the remote builder setup for aarch64

```
❯ nix flake check --accept-flake-config --allow-import-from-derivation 

ghaf on flake-parts [$] via impure (Ghaf-devshell-env) took 5m26s
```

takes 5 minutes when caches are hot.

vilvo commented:
> I'd like to add `nix flake check` to checklist (and integrate devour-flake, to check that all outputs of flake evaluable)

I think this is a good idea when `nix flake check` gets fixed. At that point, we should add it to GHA based checks like `nix flake show` https://github.com/tiiuae/ghaf/actions/runs/6981492310/workflow

vilvo commented:
973ba16 addresses most of the comments. [Compare](https://github.com/tiiuae/ghaf/compare/53f954d07718f458f02819d4c5f8af65fe7ac0e1..973ba161f4a0ae2cc2855a41f61b4a563336e95c) shows also delta from rebase with `origin/main`
31a1197 adds free text field for Testing
